core/src/main/java/org/apache/iceberg/BaseFilesTable.java

```diff
@@ -185,5 +232,60 @@ public Iterable<FileScanTask> split(long splitSize) {
     ManifestFile manifest() {
       return manifest;
     }
+
+    private List<Function<ContentFile<?>, Object>> accessors(boolean partitioned) {
```

Comment:
These two methods are hard to grasp. I think our goal is to wrap the struct we get back from reading manifests and attach metrics to it if they were projected. Let me think more on this one.

Reply:
What if you created a StructLike that would wrap the StructLike you get from reading manifests and the metrics map?
Something like this.

```
private static class SomeReasonableNameStructLike implements StructLike {
  private final StructLike fileAsStruct;
  private final Map<String, StructLike> readableMetrics;

  @Override
  public <T> T get(int pos, Class<T> javaClass) {
    if (pos < fileAsStruct.size()) {
      return fileAsStruct.get(pos, javaClass);
    } else if (pos == fileAsStruct.size()) {
      return javaClass.cast(readableMetrics);
    } else {
      throw new ...
    }
  }
```

Reply:
I gave a try but there are two issues:
1.  BaseFile::size() returns a fixed size (DataFile.getType()), regardless of the projection schema.  I guess this can be fixed, but hopefully doesnt break anything.
2. A more serious issue, BaseFile has a field called 'fileOrdinal', but it is not on the FilesTable schemas (again DataFile.getType()).  It seems its working today because its the last field, and projection on the table will never request that field.

To me a list of explicit accessors for this table is better as it decouples the table from the messiness of the underlying BaseFile/DataFile classes.  Maybe we can try to encapsulate this somehow in a StructLike class ?

Reply:
Oh, I see, let me think.

Reply:
One way to make the custom struct approach to work is to pass the projected schema, which will tell us the number of projected columns. Based on that, we can either delegate to the content file struct or return metrics.

```
public CloseableIterable<StructLike> rows() {
  if (...) {
    return CloseableIterable.transform(files(), this::wrapWithMetrics);
  } else {
    return CloseableIterable.transform(files(), file -> (StructLike) file);
  }
}
```

I don't mind using the accessor-based approach too but we can't initialize the accessor list, pick columns, etc for every row. We also can't assign them to a field as the task will be serialized. We will have to build them lazily.

Reply:
Made the suggested change.